app/scripts/lib/pairing-channel-client.js

```
@@ -0,0 +1,187 @@
+ *
+ * @export
+ * @class PairingChannelClient
+ * @extends {Model}
+ */
+export default class PairingChannelClient extends Model {
+  constructor (attrs = {}, options = {}) {
+    super(attrs, options);
+
+    this.sentryMetrics = options.sentryMetrics || Raven;
+    this._importPairingChannel = options.importPairingChannel;
+    this.set('isConnected', false);
+  }
+
+  /**
+   * Open a socket to `channelId` on `channelServerUri` with a given `channelKey`.
+   *
+   * @param {String} [channelServerUri=this.get('channelServerUri')]
+   * @param {String} [channelId=this.get('channelId')]
+   * @param {String} [channelKey=this.get('channelKey')]
+   * @returns {Promise} resolves when connected
+   */
+  open (channelServerUri = this.get('channelServerUri'),
+    channelId = this.get('channelId'),
+    channelKey = this.get('channelKey'),
+  ) {
+    return this._importPairingChannel().then((FxAccountsPairingChannel) => {
+      if (this.channel) {
+        // to avoid opening a duplicate connection, say the client is connected
+        // if a socket exists but isn't yet connected.
+        throw PairingChannelClientErrors.toError('ALREADY_CONNECTED');
+      }
+
+      if (! channelServerUri || ! channelId || ! channelKey) {
+        throw PairingChannelClientErrors.toError('INVALID_CONFIGURATION');
+      }
+
+      const psk = base64urlToArrayBuffer(channelKey);
+      return FxAccountsPairingChannel.PairingChannel.connect(channelServerUri, channelId, psk).then((channel) => {
+
+        this.channel = channel;
+        this.set('isConnected', true);
+        this.trigger('connected');
+
+        this.channel.addEventListener('message', this._messageHandler.bind(this));
+        this.channel.addEventListener('error', this._errorHandler.bind(this));
+        this.channel.addEventListener('close', this._closeHandler.bind(this));
+
+      }).catch((err) => {
+        this.sentryMetrics.captureException(err);
+        this.trigger('error', err);
```

Do you intend to swallow the connection errors here or should they be propagated out to the promise too?

---

I've looked at this a bit and so far just triggering error worked out okay.